### PR TITLE
Add parameter sweep functionality

### LINKED ISSUE #362: Sweep functionality
As this stands as a high priority to complete, I've been thinking how this can be done on a higher level first. 

Here are some general constraints, that we probably should impose to make things simpler: 
- Only one parameter to optimise at the time
- Input is always a small `preview` set of 5-7 slices. I'd go with >3 slices as it with 3d methods in place, we would like to have some adjacent neighbouring information vertically. The user can specify the previewed set but we always need to take middle 5 slices of that set. For the whole volume it will be the middle chunk. 
- The output does NOT go into the hdf5 file automatically. I'll work to resolve #283 so it is a user-controlled thing. Probably if the user enables saving, then we should warn and switch that off?  The 4D dataset could be an option though, but might not necessarily needed, if the bellow option is enabled. 
- The output saves the series of tiffs with respect to each parameter value that is being optimised in the `sweep` selection. Only the mid-slice of the previewed 5- slices set is being saved in the tiff file.  

Implementation notes: 
- we need a way how to loop the series of methods. Possibly a section should be created starting with the method which parameters are being optimised? 
- save_to_images probably can be auto-substituted but as we need just one file (mid-slice) to be saved may be we could come up with a slightly modified wrapper? The parameter value should be in the name of the tiff file created.

### PR DESCRIPTION
**Pull request by yousefmoazzam**
Attempt to fix #362

Main changes:
- add `ParamSweepWriter` + `ParamSweepReader` for writing + reading parameter sweep results using blocks
- add `Stages` to group methods that are before the sweep, in the sweep, and after the sweep
- add `ParamSweepRunner` to orchestrate the parameter sweep run
- add `SideOutputManager` to separate side output logic from runner
- refactor `DataSetBlock` and introduce three interfaces that describe the distinct sets of behaviour that a block needs to be processable by an implementor of `MethodWrapper` (see commit message of 4123373 for more info)
- add `ParamSweepBlock` to hold data during a parameter sweep run
- modify `run` command to choose between executing high-throughput or param sweep run

Things left to do:
- [x] provide a way to do parameter sweep runs from CLI
- [x] add back in the modified loader that can read the `!Sweep` and `!SweepRange` YAML "tags"
- [x] add back in the YAML checker functionality to validate YAML pipeline files with the `!Sweep` and `!SweepRange` tags
- [x] configure logger for parameter sweep run

Acceptance criteria checklist
- [ ] all tests pass
- [x] adequate documentation provided on how to use the parameter sweep functionality, what output it produces (ie, middle slices of the various sweep results) , etc
- [x] performance of parameter sweep run is acceptable

### CODE REVIEW

**Comment by yousefmoazzam** [comment]
For an example of how the logging looks so far (I've only added the basics), with the following pipeline (sweeps over 10 `center` values in the recon):
```yaml
- method: standard_tomo
  module_path: httomo.data.hdf.loaders
  parameters:
    name: tomo
    data_path: entry1/tomo_entry/data/data
    image_key_path: entry1/tomo_entry/instrument/detector/image_key
    rotation_angles:
      data_path: /entry1/tomo_entry/data/rotation_angle
    dimension: 1
    pad: 0
    preview:
      detector_y:
        start: 50
        stop: 57
- method: normalize
  module_path: httomolibgpu.prep.normalize
  parameters:
    cutoff: 10.0
    minus_log: true
    nonnegativity: false
    remove_nans: false
- method: paganin_filter_tomopy
  module_path: httomolibgpu.prep.phase
  parameters:
    pixel_size: 0.0001
    dist: 50.0
    energy: 53.0
    alpha: 0.001
- method: remove_all_stripe
  module_path: httomolibgpu.prep.stripe
  parameters:
    snr: 3.0
    la_size: 61
    sm_size: 21
    dim: 1
- method: FBP
  module_path: httomolibgpu.recon.algorithm
  save_result: False
  parameters:
    center: !SweepRange
      start: 10
      stop: 20
      step: 1
    filter_freq_cutoff: 0.6
    recon_size: null
    recon_mask_radius: null
- method: save_to_images
  module_path: httomolib.misc.images
  parameters:
    subfolder_name: images
    axis: 1
    file_format: tif
    bits: 8
    perc_range_min: 0.0
    perc_range_max: 100.0
    jpeg_quality: 95
    offset: 0
    asynchronous: true
```

the following terminal output is produced:
```
See the full log file at: /dls/tmp/twi18192/sweep-tests/08-07-2024_16_35_08_output/user.log
Loading data with shape (1801, 7, 2560)
Running data_reducer (httomolib)
Running normalize (httomolibgpu)
Running paganin_filter_tomopy (httomolibgpu)
Running remove_all_stripe (httomolibgpu)
Running FBP (httomolibgpu)
    Parameter sweep over 10 values of parameter: center
      0%|          | 0/10 [00:00<?, ?parameter/s, center=10]
     10%|#         | 1/10 [00:02<00:21,  2.36s/parameter, center=11]
     20%|##        | 2/10 [00:03<00:12,  1.51s/parameter, center=12]
     30%|###       | 3/10 [00:04<00:08,  1.25s/parameter, center=13]
     40%|####      | 4/10 [00:05<00:06,  1.12s/parameter, center=14]
     50%|#####     | 5/10 [00:06<00:05,  1.04s/parameter, center=15]
     60%|######    | 6/10 [00:06<00:03,  1.00parameter/s, center=16]
     70%|#######   | 7/10 [00:07<00:02,  1.03parameter/s, center=17]
     80%|########  | 8/10 [00:08<00:01,  1.05parameter/s, center=18]
     90%|######### | 9/10 [00:09<00:00,  1.06parameter/s, center=19]
Finished parameter sweep
Running save_to_images (httomolib)
Pipeline finished. Took 32.329s
```

Feel free to make any suggestions on tweaks/additions for improvement :slightly_smiling_face:

File: httomo/cli.py [hunk 1]
```diff
@@ -190,21 +164,50 @@ def run(
...
+                    raise ValueError(
+                        f"GPU Device not available for access. Use a GPU ID in the range: 0 to {gpu_count} (exclusive)"
+                    )
+
+                cp.cuda.Device(gpu_id).use()
+
+            httomo.globals.gpu_id = gpu_id
+
+        except ImportError:
+            pass  # silently pass and run if the CPU pipeline is given
```
Comment (team-gpu) [inline httomo/cli.py:192]: Can we factor this out into a helper function? It's a bit distracting here in this function.

**Review by team-gpu — COMMENTED** [review]
Good idea re factoring out the various block interfaces into smaller protocols, etc. 

Is there any chance to re-use more of `TaskRunner` in `ParameterSweepRunner` somehow? Although it's not too much, there is a good bit of overlap. What do you think?

File: httomo/cli.py (hunk 1 above)
Comment (yousefmoazzam) [reply httomo/cli.py:192]: Good idea, I agree that this block of code is distracting being in the top-level function of the CLI.

File: httomo/block_interfaces.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    @property
+    def is_last_in_chunk(self) -> bool:
+        """
+        Check if the current dataset is the final one for the chunk handled by the current
+        process
+        """
+        ...  # pragma: no cover
+
+
+class Block(BlockData, BlockIndexing, BlockTransfer, Protocol):
```
Comment (team-gpu) [inline httomo/block_interfaces.py:130]: There's on need for Protocol as base class here, as all 3 others are already Protocols

**Review by team-gpu — APPROVED** [review]
Good to go for me.

File: httomo/cli.py (hunk 1 above)
Comment (yousefmoazzam) [reply httomo/cli.py]: Done in c4ca5ff

File: httomo/block_interfaces.py (hunk 2 above)
Comment (yousefmoazzam) [reply httomo/block_interfaces.py]: Cheers, didn't know that it could be left out, done in af0ff1b

**Pull request merged**